### PR TITLE
luamodel update ...

### PR DESCRIPTION
**Pull request by mjhmilla**
A modified version of the lua updates: rbdl does not include the lastest and very large overhaul of the constraints system that is in rbdl-orb. Both versions rbdl and rbdl-orb will be able to load the same lua files, but rbdl is missing all of the generic functions to work with large constraint sets that are in rbdl-orb. You can see this clearly if you have a look at the include/rbdl folders of rbdl and rbdl-orb as these files are missing in rbdl:

-Constrant.h : the base class
-Constraint_Contact : the contact constraint that extends from Constraint
-Constraint_Loop : the loop constraint that extends from Constraint

This overhaul was done for two reasons:

1. To provide generic functions to get all of the relevant quantities (errors and forces): many people were writing their own code to do these things and screwing it up. With the current constraint update a constraint author (e.g. an expert) writes these functions once and others then use them: far fewer screw-ups result.
2. To provide a number of human-friendly access functions to get and manipulate specific constraints: this is absolutely necessary for the large complex models that I work with.
3. To provide a fast an easily extensible base class. A simple caching system was added (see line 34 Constraint.h) and the Constraint base class functions write directly into their specific part of the mass matrix or the right hand side: no new memory is used at run time.

This is big. We need to discuss how you plan to move forward. In general I am unhappy that we have two 'main' rbdl repositories: rbdl and rbdl-orb. It would be far better, I think, if we were able to resolve our issues and work on a single repository.

As far as I can see there are at least 3 issues on the table:

1. rbdl-orb has error handling so that rbdl can be used without crashing other programs. 
2. rbdl-orb ditched simple math: it lacks some of the functions of Eigen that we made use of. We could perhaps find a way to compromise on this.
3. rbdl-orb has a drastically different constraint system. This is a giant compatibility issue for anyone who's using the C++ api.

--
Specific details about this pull request ...
--
1. New structures and functions added to luamodel

- Point & LuaModelReadPoints
- LocalFrame & LuaModelReadLocalFrames
- MotionCaptureMarker & LuaModelReadMotionCaptureMarkers

2. The function LuaModelReadFromFileWithConstraints has been updated so that Contact and Loop constraints can take advantage of a more compact syntax by using the 'point' and 'local_frame' sets.

3. The test code in addons/luamodel/tests/testLuaModel.cc has been greatly expanded: the read in model is compared against the known values for almost every exposed field. The only field currently missing from the test is the local transform between a body and the point at which a Contact/Loop constraints are applied.

4. A small tweak has been made to AddContactConstraint and AddLoopConstraint: subsequent constraints are put in the same group now if they are the same type, apply to the same bodies and local points, and have the same user defined ids.

5. The sample lua models have been commented a bit. Between reading the doxygen in Constraints.h and the lua comments an end user of RBDL should be able to figure out what's going on.

### CODE REVIEW

File: addons/luamodel/luamodel.h [hunk 1]
```diff
@@ -261,14 +262,69 @@ namespace Addons {
...
+  @param pointSet: an empty std::vector of Point structure
+  @param verbose: information will be printed to the command window if this
+                  is set to true.
+
+*/
+RBDL_DLLAPI
+bool LuaModelReadMotionCaptureMarkers (
+      const char* filename,
+      const Model* model,
+      std::vector<Point> &markerSet,
```
Comment (martinfelis) [inline addons/luamodel/luamodel.h:288]: naming convention: use `marker_set` instead of `markerSet`

File: addons/luamodel/luatypes.h [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+{
+  MotionCaptureMarker result = default_value;
+
+  if (stackQueryValue()) {
+    LuaTable marker_table = LuaTable::fromLuaState (luaTable->L);
+
+    result.name         = marker_table["name"].get<std::string>();
+    result.point_local  = marker_table["point"]
+                            .getDefault<RigidBodyDynamics::Math::Vector3d>(
+                              result.point_local);
```
Comment (martinfelis) [inline addons/luamodel/luatypes.h:353]: Seems to be missing the `body` parameter.

File: addons/luamodel/luatypes.h [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+{
+  LocalFrame result = default_value;
+
+  if (stackQueryValue()) {
+    LuaTable local_frame_table = LuaTable::fromLuaState (luaTable->L);
+
+    result.name      = local_frame_table["name"].get<std::string>();
+    result.body_name = local_frame_table["body"].get<std::string>();
+    result.r = local_frame_table["r"]
+                .getDefault<RigidBodyDynamics::Math::Vector3d>(result.r);
```
Comment (martinfelis) [inline addons/luamodel/luatypes.h:375]: If in Lua `frame.r` is not given then `result.r` will be undefined. The `getDefault()` function assumes that you provide the value that should be used if the value is not found in the Lua table. In this case you likely want
```
.getDefault<RigidBodyDynamics::Math::Vector3d>(Vector3d::Zero());
```
this also holds for `result.E` below where you need to use `Matrix3d::Identity()` as default value.

File: addons/luamodel/luamodel.h (hunk 1 above)
Comment (mjhmilla) [reply addons/luamodel/luamodel.h]: All of the camel-case names have been updated to snake-case (including the doxygen).

File: addons/luamodel/luatypes.h (hunk 2 above)
Comment (mjhmilla) [reply addons/luamodel/luatypes.h]: Yikes! Included now.

File: addons/luamodel/luatypes.h (hunk 3 above)
Comment (mjhmilla) [reply addons/luamodel/luatypes.h]: I've made this change. However, I think the problem might have been my understanding. Is this correct?

- When the line 'LocalFrame result = default_value' is executed the default constructor for LocalFrame is called and thus the r and E fields have values.
- By passing result.r and result.E into the 'getDefault' command the values for r & E set in the default constructor are used (in this case a Vector3d and Matrix3d of nans).

Does this not work?

**Pull request closed without merging**